libbeat/processors/dns/cache.go

```diff
@@ -0,0 +1,194 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package dns
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+type ptrRecord struct {
+	host    string
+	expires time.Time
+}
+
+func (r ptrRecord) IsExpired(now time.Time) bool {
+	return now.After(r.expires)
+}
+
+type ptrCache struct {
+	data    map[string]ptrRecord
+	maxSize int
+}
+
+func (c *ptrCache) set(now time.Time, key string, ptr *PTR) {
+	if len(c.data) >= c.maxSize {
+		c.evict()
+	}
+
+	c.data[key] = ptrRecord{
+		host:    ptr.Host,
+		expires: now.Add(time.Duration(ptr.TTL) * time.Second),
+	}
+}
+
+func (c *ptrCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *ptrCache) get(now time.Time, key string) *PTR {
+	r, found := c.data[key]
+	if found && !r.IsExpired(now) {
+		return &PTR{r.host, uint32(r.expires.Sub(now) / time.Second)}
+	}
+	return nil
+}
+
+type failureRecord struct {
+	error
+	expires time.Time
+}
+
+func (r failureRecord) IsExpired(now time.Time) bool {
+	return now.After(r.expires)
+}
+
+type failureCache struct {
+	data       map[string]failureRecord
+	maxSize    int
+	failureTTL time.Duration
+	stats      cacheStats
+}
+
+func (c *failureCache) set(now time.Time, key string, err error) {
+	if len(c.data) >= c.maxSize {
+		c.evict()
+	}
+
+	c.data[key] = failureRecord{
+		error:   err,
+		expires: now.Add(c.failureTTL),
+	}
+}
+
+func (c *failureCache) evict() {
+	var key string
+	for k := range c.data {
+		key = k
+		break
+	}
+	delete(c.data, key)
+}
+
+func (c *failureCache) get(now time.Time, key string) error {
```

Comment:
My read here is that the tradeoff is between:

1. Dual cache impl, one per type
2. 'Generic' cache impl, with entries of `(key,success|failure)` (probably using two pointers, only one of which is non-null

where the first option is more memory efficient at the cost of duplicated code, while the second is a little less memory efficient but more DRY approach (though there is some extra complexity in methods since they have to determine the type they're working with).

I'm neutral on the choice. It's a tradeoff, and the difference in terms of both isn't very large since 1. the code is simple. 2. finding the hash code of a short string is fast, but I thought I'd just bring this up because I'm curious if there's general guidance on this issue.

In previous PL's I've used professionally the answer is generics. Not having them forces this choice, so I'd like to hear from more experience go programmers how this is handled.

Reply:
Makes sense, thanks for the feedback.